core/src/main/scala/kafka/log/Log.scala

```
@@ -1935,4 +2040,122 @@ object Log {
+      if (validBatches.size == 0)
+        return new CopyResult(None, overflowOffset)
+
+      // read all the valid batches
+      val lastValidBatch = validBatches.last
+      val validRecords = records.slice(0, lastValidBatch.position + lastValidBatch.sizeInBytes)
+      require(readBuffer.capacity >= validRecords.sizeInBytes)
+      readBuffer.clear()
+      readBuffer.limit(validRecords.sizeInBytes)
+      validRecords.readInto(readBuffer, 0)
+
+      for (batch <- validBatches) {
+        if (batch.maxTimestamp > maxTimestamp) {
+          maxTimestamp = batch.maxTimestamp
+          offsetOfMaxTimestamp = batch.lastOffset
+        }
+
+        maxOffset = batch.lastOffset
+        bytesRead += batch.sizeInBytes
+      }
+
+      // append valid batches into the segment
+      require(bytesRead > 0, "no valid bytes found during split")
+      segment.append(maxOffset, maxTimestamp, offsetOfMaxTimestamp, MemoryRecords.readableRecords(readBuffer))
+      readBuffer.clear()
+      log.info(s"Appended messages till $maxOffset to segment $segment during split")
+
+      return new CopyResult(Some(bytesRead), overflowOffset)
+    }
+
+    log.info(s"Splitting segment $segment in log $log")
+    newSegments += LogCleaner.createNewCleanedSegment(log, segment.baseOffset)
+    while (position < sourceRecords.sizeInBytes) {
+      val currentSegment = newSegments.last
+
+      // grow buffers if needed
+      val firstBatch = sourceRecords.slice(position, Int.MaxValue).batches.asScala.head
+      if (firstBatch.sizeInBytes > readBuffer.capacity)
+        readBuffer = ByteBuffer.allocate(firstBatch.sizeInBytes)
+
+      // get records we want to copy and copy them into the new segment
+      val recordsToCopy = sourceRecords.slice(position, readBuffer.capacity)
+      val copyResult = copyRecordsToSegment(recordsToCopy, currentSegment, readBuffer)
+      require(copyResult.bytesRead.isDefined, "could not copy records into new segment")
```

The types are a little confusing. If we are going to raise an error if `CopyResult.bytesRead` is not defined, why let it be an option in the first place? We could raise an exception directly from `copyRecordsToSegment` and skip this requirement.

---

Hmm, the current expectation does not seem to be correct. I think we could be in a situation where the first batch passed into `copyRecordsToSegment` causes offset overflow and we need to create a new segment for it. I updated the implementation to take this into account.